plutus-ir/src/Language/PlutusIR/Parser.hs

```diff
@@ -285,49 +289,51 @@ binding =  inParens $
     <|> (reservedWord "typebind" >> TypeBind <$> getPosition <*> tyVarDecl <*> typ)
     <|> (reservedWord "datatypebind" >> DatatypeBind <$> getPosition <*> datatype)
 
-absTerm :: Parser (Term TyName Name SourcePos)
-absTerm = TyAbs <$> reservedWord "abs" <*> tyVar <*> kind <*> term
+-- A small type wrapper for parsers that are parametric in the type of term they parse
+type Parametric = forall term. PIR.TermLike term TyName Name => Parser (term SourcePos) -> Parser (term SourcePos)
 
-lamTerm :: Parser (Term TyName Name SourcePos)
-lamTerm = LamAbs <$> reservedWord "lam" <*> name <*> typ <*> term
+absTerm :: Parametric
+absTerm tm = PIR.tyAbs <$> reservedWord "abs" <*> tyVar <*> kind <*> tm
 
-conTerm :: Parser (Term TyName Name SourcePos)
-conTerm = Constant <$> reservedWord "con" <*> constant
+lamTerm :: Parametric
+lamTerm tm = PIR.lamAbs <$> reservedWord "lam" <*> name <*> typ <*> tm
 
-iwrapTerm :: Parser (Term TyName Name SourcePos)
-iwrapTerm = IWrap <$> reservedWord "iwrap" <*> typ <*> typ <*> term
+conTerm :: Parametric
+conTerm _tm = PIR.constant <$> reservedWord "con" <*> constant
 
-builtinTerm :: Parser (Term TyName Name SourcePos)
-builtinTerm = Builtin <$> reservedWord "builtin" <*> builtinVar
+iwrapTerm :: Parametric
+iwrapTerm tm = PIR.iWrap <$> reservedWord "iwrap" <*> typ <*> typ <*> tm
 
-unwrapTerm :: Parser (Term TyName Name SourcePos)
-unwrapTerm = Unwrap <$> reservedWord "unwrap" <*> term
+builtinTerm :: Parametric
+builtinTerm _term = PIR.builtin <$> reservedWord "builtin" <*> builtinVar
 
-errorTerm :: Parser (Term TyName Name SourcePos)
-errorTerm = Error <$> reservedWord "error" <*> typ
+unwrapTerm :: Parametric
+unwrapTerm tm = PIR.unwrap <$> reservedWord "unwrap" <*> tm
+
+errorTerm :: Parametric
+errorTerm _tm = PIR.error <$> reservedWord "error" <*> typ
 
 letTerm :: Parser (Term TyName Name SourcePos)
 letTerm = Let <$> reservedWord "let" <*> recursivity <*> some (try binding) <*> term
 
-appTerm :: Parser (Term TyName Name SourcePos)
-appTerm = do
-    pos  <- getPosition
-    fn   <- term
-    args <- some term
-    pure $ foldl' (Apply pos) fn args
+appTerm :: Parametric
+appTerm tm = PIR.mkIterApp <$> getPosition <*> tm <*> some tm
 
-tyInstTerm :: Parser (Term TyName Name SourcePos)
-tyInstTerm = do
-    pos  <- getPosition
-    fn   <- term
-    args <- some typ
-    pure $ foldl' (TyInst pos) fn args
+tyInstTerm :: Parametric
+tyInstTerm tm = PIR.mkIterInst <$> getPosition <*> tm <*> some typ
+
+term' :: Parametric
+term' other = (var >>= (\n -> return $ PIR.var (nameAttribute n) n))
```

Comment:
Does this typecheck? :thinking:

Reply:
Oh I see, your `Parametric` takes an argument parser too. That's a bit confusing...

Reply:
Yes! (I've compiled and tested it) What about it looks wrong to you?

Reply:
> Oh I see, your `Parametric` takes an argument parser too. That's a bit confusing...

I agree it's a bit ugly; it's mostly there to avoid some typing - shouldn't be much of an issue since neither `Parametric` nor `term'` are exported, so it's not really visible from the outside.